rfcs/20190117-tf-module.md

```diff
@@ -0,0 +1,634 @@
+long as the super constructor has run).
+
+Variable reuse should be achieved using standard object oriented programming
+(e.g. assigning to object properties). `tf.Module` doesn't do anything behind the
+scenes to help (or hinder) variable reuse.
+
+We considered including a default implementation of `__call__` since lazy
+building is extremely common (e.g. constructing model parameters the first time
+an input is observed), however we backtracked on this for a number of reasons:
+
+-   Even among the authors there existed different preferences for how to enable
+    this pattern, it seemed unlikely we could make a choice which would satisfy
+    the community.
+-   Not all modules want a `__call__` method (e.g. an `Optimizer`) and we got
+    strong feedback that it is annoying to ignore unused methods inherited from
+    your base class.
+-   Encouraging a lazy pattern might make it harder to create modules with
+    multiple forward methods (e.g. a VAE module might be implemented with an
+    `encode` and `decode` method with no `__call__`).
+
+As such we leave the problem of how to enable lazy parameter construction to
+library authors who build on `tf.Module`s.
+
+### Checkpointing
+
+`tf.Module` will extend `tf.checkpoint.AutoCheckpointable` which automatically
+tracks variables which are object properties and enables restore on create etc.
+Users who want to learn more about checkpointing should look into this type.
+
+It is worth noting that object structure (e.g. the names of object parameters)
+are encoded into checkpoints, so library authors should take care to lay out
+their objects in a consistent way and defend their structure. `tf.Module` is
+careful to not add any load to the user here by not imposing any requirements on
+how you lay out your internal state.
+
+We considered imposing a constraint that all `tf.Variable`s were created with some
+utility (e.g. `self.create_variable(name, shape)`) and storing these in a
+dictionary of variable name to instance. While this would simplify the checkpointing
+story (since the structure of TensorFlow state would only depend on variable
+names) internal users felt that it was overly restrictive. Additionally we would
+require functions to query this storage (e.g. `get_variable('name')` and remove
+items from it). Library authors may want to consider this pattern if it makes
+sense for their users.
+
+### `tf.Variable` tracking
+
+It is very convenient to be able to collect all the variables for a given module
+and all of its dependencies. We achieve this in `tf.Module` using reflection to
+walk object properties. When walking if we encounter a nested structure (e.g. a
+`list`, `dict` etc) we will flatten it using `nest.flatten(value)` and evaluate
+each leaf. This means variables nested in properties will be found (e.g.
+variables in `list`s or `tuple`s on the module).
+
+This "walk" function is used to provide the following properties:
```

Comment:
Are there properties computed once per instance, or are they fully dynamic?

Reply:
Fully dynamic, trying to reason about when to invalidate the caches would be non-trivial.

Reply:
I'm thinking the performance cost of having this dynamic should be well documented so that users know to cache the results of these calls if using eager and good performance is required. Though, if everything works seamlessly with `tf.function`, then maybe this wouldn't be necessary and encouraging users to use `tf.function` for better performance would be sufficient.

Reply:
Agreed re documentation, I'll add a note in the docstrings.

For a 100 linear layer sequential it takes 2ms to collect all 200 parameters (at least on my workstation). This could be worse for modules with deep hierarchies or large nests.

Assuming the call to `trainable_variables` (or whatever) is inside the `tf.function` this cost is only paid when the function is traced (~once).